pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,43 +88,84 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
+// SaveAlertInstances saves all the provided alert instances to the store. It
+// writes transactions with a maximum count of 400 rows. Larger writes are
+// split into multiple transactions.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	// SQLite has a limit of 999 variables per write. We use 9 per row, so we split up our writes into separate upsert statements.
+	err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		keysPerRow := 9
```

Comment:
Could we make this variable the length of `[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"}` such that when the slice is updated so is `keysPerRow`?